datajoint/external.py

```diff
@@ -21,63 +21,65 @@ class ExternalTable(Table):
     The table tracking externally stored objects.
     Declare as ExternalTable(connection, database)
     """
-    def __init__(self, arg, database=None):
-        if isinstance(arg, ExternalTable):
-            super().__init__(arg)
-            # copy constructor
-            self.database = arg.database
-            self._connection = arg._connection
+    def __init__(self, connection, store=None, database=None):
+
+        # copy constructor -- all QueryExpressions must provide
+        if isinstance(connection, ExternalTable):
+            other = connection   # the first argument is interpreted as the other object
+            super().__init__(other)
+            self.store = other.store
+            self.spec = other.spec
+            self.database = other.database
+            self._connection = other._connection
             return
+
+        # nominal constructor
         super().__init__()
+        self.store = store
+        self.spec = config.get_store_spec(store)
         self.database = database
-        self._connection = arg
+        self._connection = connection
         if not self.is_declared:
             self.declare()
 
     @property
     def definition(self):
         return """
         # external storage tracking
-        hash  : {hash_data_type}  # the hash of stored object + store name
+        hash  : uuid
         ---
         size      :bigint unsigned   # size of object in bytes
         timestamp=CURRENT_TIMESTAMP  :timestamp   # automatic timestamp
-        """.format(hash_data_type=HASH_DATA_TYPE)
+        """
 
     @property
     def table_name(self):
-        return '~external'
+        return '{external_table_root}_{store}'.format(external_table_root=EXTERNAL_TABLE_ROOT, store=self.store)
 
-    def put(self, store, blob):
+    def put(self, blob):
         """
         put an object in external store
         """
-        store = ''.join(store.split('-')[1:])
-        spec = config.get_store_spec(store)
-        blob_hash = long_hash(blob) + store
-        if spec['protocol'] == 'file':
-            folder = os.path.join(spec['location'], self.database, *subfold(blob_hash, spec['subfolding']))
-            full_path = os.path.join(folder, blob_hash)
+        blob_hash = uuid_from_buffer(blob)
+        if self.spec['protocol'] == 'file':
+            folder = os.path.join(self.spec['location'], self.database, *subfold(blob_hash.hex, self.spec['subfolding']))
+            full_path = os.path.join(folder, blob_hash.hex)
```

Comment:
clearly, as this line illustrates, not a complicated conversion, however, would it be preferable to save directly to the uuid string so hash:filename is 1:1? (but still with subpaths path splitting?)

Reply:
It uses `binary(16)` to store UUIDs.  When you view the external table in DataJoint, the UUIDs show as UUID strings. If you need to view them in SQL, you can do something like `SELECT hex(hash) FROM external_local`